### PR TITLE
2918 replace json macro

### LINKED ISSUE #2918: Replace uses of json! with OAuthTokenRequest when calling make_oauth_token_request.
As part of #2896, a new `struct` `OAuthTokenRequest` was introduced. 

This _should_ used for multiple other places in `http_client.rs`— callers of `make_oauth_token_request`— which currently using a raw `json!` macro.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1066)
┆Sprint: Backlog

### PR DESCRIPTION
**Pull request by yabirgb**
Fixes #2918 .

I've modified the previously introduced `OAauthTokenRequest` introducing an `enum` as recommended making the distinction for `RefreshToken` and `Code`. 

- Modified `OAauthTokenRequest` to make it a serializable enum
- Modified the code that make use of `make_oauth_token_request`

### Pull Request checklist ###

- This change doesn't need to be added to the changes list since only refactors the code.
- There is no new functionalities added to the code so I don't see the need for new tests

- [ ] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/fxa-client/src/http_client.rs [hunk 1]
```diff
@@ -661,14 +662,31 @@ pub struct DeviceResponseCommon {
...
-    pub grant_type: String,
-    pub refresh_token: String,
-    pub scope: String,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub ttl: Option<u64>,
+#[serde(untagged)]
+enum OAauthTokenRequest {
+    RefreshToken {
+        client_id: String,
+        grant_type: String,
```
Comment (eoger) [inline components/fxa-client/src/http_client.rs:674]: I think I gave you un-helpful advice earlier, you should probably use [Internally tagged](https://serde.rs/enum-representations.html#internally-tagged) enums instead, and use `grant_type` as the key. I would guess you'd need to `#[serde(rename` the different enum variants to match the excepted value in `grant_type`?

File: components/fxa-client/src/http_client.rs [hunk 2]
```diff
@@ -661,14 +662,31 @@ pub struct DeviceResponseCommon {
...
-pub struct OAuthTokenRequest {
-    pub client_id: String,
-    pub grant_type: String,
-    pub refresh_token: String,
-    pub scope: String,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub ttl: Option<u64>,
+#[serde(untagged)]
+enum OAauthTokenRequest {
+    RefreshToken {
```
Comment (eoger) [inline components/fxa-client/src/http_client.rs:672]: Maybe to make it clear, we could rename these variants `UsingRefreshToken`, `UsingCode`?

File: components/fxa-client/src/http_client.rs [hunk 3]
```diff
@@ -661,14 +662,31 @@ pub struct DeviceResponseCommon {
     pub push_endpoint_expired: bool,
 }
 
+// We modele the OAuthTokenRequest according to the up to date
```
Comment (eoger) [inline components/fxa-client/src/http_client.rs:665]:
```suggestion
// We model the OAuthTokenRequest according to the up to date
```

File: components/fxa-client/src/http_client.rs (hunk 2 above)
Comment (yabirgb) [reply components/fxa-client/src/http_client.rs:672]: Yes seems a good name. I was using `WithRefreshToken` but the syntax checker advised me not to

File: components/fxa-client/src/http_client.rs (hunk 1 above)
Comment (yabirgb) [reply components/fxa-client/src/http_client.rs]: I believe I understand what you mean. I've updated the code

File: components/fxa-client/src/http_client.rs [hunk 4]
```diff
@@ -661,14 +660,30 @@ pub struct DeviceResponseCommon {
...
+        ttl: Option<u64>,
+    },
+    #[serde(rename = "authorization_code")]
+    UsingCode {
+        client_id: String,
+        code: String,
+        code_verifier: String,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        ttl: Option<u64>,
+    },
```
Comment (eoger) [inline components/fxa-client/src/http_client.rs:686]: Do you mind adding these tests at the end of the file:

```
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[allow(non_snake_case)]
    fn check_OAauthTokenRequest_serialization() {
        // Ensure OAauthTokenRequest serializes to what the server expects.
        let using_code = OAauthTokenRequest::UsingCode {
            code: "foo".to_owned(),
            client_id: "bar".to_owned(),
            code_verifier: "bobo".to_owned(),
            ttl: None,
        };
        assert_eq!("{\"grant_type\":\"authorization_code\",\"client_id\":\"bar\",\"code\":\"foo\",\"code_verifier\":\"bobo\"}", serde_json::to_string(&using_code).unwrap());
        let using_code = OAauthTokenRequest::UsingRefreshToken {
            client_id: "bar".to_owned(),
            refresh_token: "foo".to_owned(),
            scope: Some("bobo".to_owned()),
            ttl: Some(123),
        };
        assert_eq!("{\"grant_type\":\"refresh_token\",\"client_id\":\"bar\",\"refresh_token\":\"foo\",\"scope\":\"bobo\",\"ttl\":123}", serde_json::to_string(&using_code).unwrap());
    }
}
```

Had to convince myself serde_json did the right thing :)

**Comment by eoger** [comment]
Also you will need to [squash](https://github.com/wprig/wprig/wiki/How-to-squash-commits) and sign your commits.

File: components/fxa-client/src/http_client.rs (hunk 4 above)
Comment (yabirgb) [reply components/fxa-client/src/http_client.rs:686]: sure

**Comment by eoger** [comment]
Thank you so much!

**Comment by yabirgb** [comment]
@eoger Thank you for your help! This was nice to get started with this project and let me learn more about serde and rust (and git too) :smile:

**Pull request merged**